diff --git a/include/GameConfiguration.h b/include/GameConfiguration.h
new file mode 100644
index 0000000..bbb5650
--- /dev/null
+++ b/include/GameConfiguration.h
@@ -0,0 +1,25 @@
+#pragma once
+
+#include "Prelude.h"
+
+constexpr int SCREEN_WIDTH = 800;
+constexpr int SCREEN_HEIGHT = 600;
+constexpr int RANGE_X = 801;
+constexpr int RANGE_Y = 601;
+
+constexpr int ENTITY_HEALTH = 100;
+constexpr int PLAYER_SPRITE_DEPTH = 10;
+constexpr float ENTITY_ANIMATION_SPEED = 15.0f;
+constexpr float ENEMY_MOVEMENT_SPEED = 90.0f;
+constexpr geometry::Vec2 PLAYER_POSITION{400, 300};
+constexpr geometry::Vec2 ENTITY_SPEED_VECTOR{1, 0};
+
+constexpr int INITIAL_SCORE = 0;
+constexpr int FONT_SIZE = 30;
+constexpr int TEXT_SPRITE_DEPTH = -7;
+constexpr float HEALTHBAR_HEALTH = 100.0f;
+constexpr float ENV_ANIMATION_SPEED = 1.0f;
+constexpr geometry::Vec2 HEALTHBAR_SPRITE_SCALE{0.5, 0.5};
+constexpr SDL_Color DEFAULT_TEXT_COLOR{238, 38, 38};
+constexpr SDL_Color SPECIAL_TEXT_COLOR{0, 0, 0};
+
